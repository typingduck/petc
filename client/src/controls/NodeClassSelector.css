#petc-node-select-control {
  background-color: var(--modal-bg);
  border-radius: var(--modal-border-radius);
  border: var(--modal-border);
  border-right-width: var(--modal-border-wide);
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  border-left: none;
  box-shadow: var(--modal-box-shadow);
  left: 0;
  min-height: 500px;
  padding: 0;
  position: absolute;
  top: 20px;
  z-index: var(--z-index-modal);
}

#petc-node-select-control ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

#petc-node-select-control li {
  align-items: center;
  border-bottom: 1px solid #555;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 2px;
  opacity: 0.8;
  padding: 10px;
}


#petc-node-select-control li.selected {
  background-color: white;
}
