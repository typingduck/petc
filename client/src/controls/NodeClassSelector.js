import React from 'react'
import Draggable from 'react-draggable'

import './NodeClassSelector.css'
import {createNode} from '../model/model'

class NodeClassSelector extends React.Component {
  constructor (props) {
    super(props)
    this._mouseDown = false

    this.dragHandleStop = this.dragHandleStop.bind(this)
    this.onMouseDown = this.onMouseDown.bind(this)
    this.createNodeClassView = this.createNodeClassView.bind(this)
  }

  dragHandleStop (ev, data) {
    const nodeClass = Array.from(data.node.classList)
      .filter(s => !s.includes('react-draggable'))
      .filter(s => !s.includes('default'))
      .filter(s => !s.includes('petc-node'))
      .join(' ')
      .trim()

    if (data.x < 60) {
      // treat as a click
    } else {
      this._mouseDown = false
      let borderWidth = parseInt(ev.target.style.borderWidth, 10)
      if (isNaN(borderWidth)) {
        borderWidth = 10  // .petc-node default border width in css
      }
      const kanvasPageOffset = 10
      let x = ev.pageX - (kanvasPageOffset + borderWidth + ev.offsetX)
      let y = ev.pageY - (kanvasPageOffset + borderWidth + ev.offsetY)
      this.props.addNode(createNode(x, y, nodeClass))
    }
  }

  classSelected (nodeClass) {
    return this.props.controls.selectedNodeClass === nodeClass
  }

  onMouseDown () {
    // work around for distinguishing click from drag
    this._mouseDown = true
  }

  handleClassSelect (nodeClass) {
    if (this._mouseDown) {
      if (this.classSelected(nodeClass)) {
        this.props.selectNodeClass(null)
      } else {
        this.props.selectNodeClass(nodeClass)
      }
      this._mouseDown = false
    }
  }

  liClassName (className) {
    return this.classSelected(className) ? 'selected' : null
  }

  createNodeClassView (className, style) {
    let fullClassName = 'petc-node ' + className
    return (<li key={className} className={this.liClassName(className)} onMouseDown={this.onMouseDown} onMouseUp={() => this.handleClassSelect(className)}>
      <Draggable position={{x: 0, y: 0}} onStop={this.dragHandleStop}>
        <div className={fullClassName} style={style} />
      </Draggable>
      <div>{className}</div>
    </li>)
  }

  render () {
    const nodeClasses = Object.assign(
      { default: {} },
      (this.props.doc.style && this.props.doc.style.nodes) || {}
    )

    if (this.props.controls.isNodeMode) {
      return (
        <div id='petc-node-select-control'>
          <ul>
            {Object.entries(nodeClasses).map(([className, style]) =>
              this.createNodeClassView(className, style)
            )}
          </ul>
        </div>
      )
    } else {
      return null
    }
  }
}

export default NodeClassSelector
