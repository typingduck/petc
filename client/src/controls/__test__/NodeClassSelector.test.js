/* global beforeEach expect it jest */
import React from 'react'
import Renderer from 'react-test-renderer'

import NodeClassSelector from '../NodeClassSelector'

jest.mock('uuid/v4', () => () => 'u-u-i-d')

let props

beforeEach(() => {
  props = {
    controls: {
      isNodeMode: true,
      isEdgeMode: false,
      selectedNodeClass: null
    },
    doc: {
      style: {
        nodes: {
          'class-red': {
            color: 'red'
          },
          'class-blue': {
            color: 'blue'
          }
        }
      }
    },
    addNode: jest.fn(),
    selectNodeClass: jest.fn()
  }
})

it('hidden unless node view mode', () => {
  const props = {
    controls: {
      isNodeMode: false
    },
    doc: {}
  }
  const component = Renderer.create(<NodeClassSelector {...props} />)
  expect(component.toJSON()).toBeNull()
})

it('visible when node view mode', () => {
  const props = {
    controls: {
      isNodeMode: true
    },
    doc: {}
  }
  const component = Renderer.create(<NodeClassSelector {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('shows custom node styles', () => {
  const component = Renderer.create(<NodeClassSelector {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('highlights selected class', () => {
  props.controls.selectedNodeClass = 'class-blue'

  const component = Renderer.create(<NodeClassSelector {...props} />)
  const ul = component.toJSON().children[0]

  const defaultNode = nodeFromUl(ul, 0)
  expect(defaultNode.props.className).toBeNull()

  const redNode = nodeFromUl(ul, 1)
  expect(redNode.props.className).toBeNull()

  const blueNode = nodeFromUl(ul, 2)
  expect(blueNode.props.className).toEqual(expect.stringContaining('selected'))
})

it('click to select node', () => {
  const component = Renderer.create(<NodeClassSelector {...props} />)
  const ul = component.toJSON().children[0]

  // drag a node (only a short distance (so should be a click)
  const redNode = nodeFromUl(ul, 1)
  simulateClickNode(component, redNode)
  expect(props.selectNodeClass).toHaveBeenCalledWith('class-red')
  props.controls.selectedNodeClass = 'class-red'

  const blueNode = nodeFromUl(ul, 2)
  simulateClickNode(component, blueNode)
  expect(props.selectNodeClass).toHaveBeenCalledWith('class-blue')
  props.controls.selectedNodeClass = 'class-blue'

  // should deselect on second click
  simulateClickNode(component, blueNode)
  expect(props.selectNodeClass).toHaveBeenCalledWith(null)
})

it('drag to add node', () => {
  const component = Renderer.create(<NodeClassSelector {...props} />)
  const ul = component.toJSON().children[0]

  // drag a node onto kanvas
  const redNode = nodeFromUl(ul, 1)
  simulateDragNode(component, redNode, 100, 100)
  expect(props.addNode).toHaveBeenCalledWith(
    {'className': 'class-red', 'id': 'u-u-i-d', 'x': 30, 'y': 30}
  )
})

function nodeFromUl (ul, index) {
  return ul.children[index]
}

function simulateClickNode (component, nodeLi) {
  simulateDragNode(component, nodeLi, 0, 0)
}

function simulateDragNode (component, nodeLi, dx, dy) {
  nodeLi.props.onMouseDown()
  const node = nodeLi.children[0]
  component.getInstance().dragHandleStop(
    { target: { style: {} }, pageX: dx, pageY: dy, offsetX: 50, offsetY: 50 },
    { node: { classList: node.props.className.split(' ') }, x: dx, y: dy }
  )
  nodeLi.props.onMouseUp()
}
