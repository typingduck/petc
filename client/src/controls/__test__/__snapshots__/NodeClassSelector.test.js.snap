// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows custom node styles 1`] = `
<div
  id="petc-node-select-control"
>
  <ul>
    <li
      className={null}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="petc-node default react-draggable"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "touchAction": "none",
            "transform": "translate(0px,0px)",
          }
        }
        transform={null}
      />
      <div>
        default
      </div>
    </li>
    <li
      className={null}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="petc-node class-red react-draggable"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "color": "red",
            "touchAction": "none",
            "transform": "translate(0px,0px)",
          }
        }
        transform={null}
      />
      <div>
        class-red
      </div>
    </li>
    <li
      className={null}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="petc-node class-blue react-draggable"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "color": "blue",
            "touchAction": "none",
            "transform": "translate(0px,0px)",
          }
        }
        transform={null}
      />
      <div>
        class-blue
      </div>
    </li>
  </ul>
</div>
`;

exports[`visible when node view mode 1`] = `
<div
  id="petc-node-select-control"
>
  <ul>
    <li
      className={null}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="petc-node default react-draggable"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "touchAction": "none",
            "transform": "translate(0px,0px)",
          }
        }
        transform={null}
      />
      <div>
        default
      </div>
    </li>
  </ul>
</div>
`;
